File: generate_stats.py
import random
import numpy as np

# formula for ability score --> ability mod
def calc_mod_from_score(ability_score):
    ability_mod = (ability_score-10)//2
    return ability_mod

# abbreviated ability scores
ability_scores = ['STR', 'DEX', 'CON', 'INT', 'WIS', 'CHA']

# race bonuses as np arrays, case and spelling must match race generation
race_bonuses = {
    'Dragonborn': np.array([0, 2, 0, 0, 0, 1]),
    'Hill Dwarf': np.array([0, 0, 2, 0, 1, 0]),
    'Mountain Dwarf': np.array([2, 0, 2, 0, 0, 0]),
    'Drow Elf': np.array([0, 2, 0, 0, 0, 1]),
    'High Elf': np.array([0, 2, 0, 1, 0, 0]),
    'Wood Elf': np.array([0, 2, 0, 0, 1, 0]),
    'Forest Gnome': np.array([0, 1, 0, 2, 0, 0]),
    'Rock Gnome': np.array([0, 0, 1, 2, 0, 0]),
    'Half-Elf': np.array([0, 0, 0, 0, 0, 2]), # two additional +1's
    'Lightfoot Halfling': np.array([0, 2, 0, 0, 0, 1]),
    'Stout Halfling': np.array([0, 2, 1 , 0, 0, 0]),
    'Half-Orc': np.array([2, 0, 1, 0, 0, 0]),
    'Human': np.array([1, 1, 1, 1, 1, 1]),
    'Variant Human': np.array([0, 0, 0, 0, 0, 0]), # choose two +1's
    'Tiefling': np.array([0, 0, 0, 1, 0, 2]),
    'Orc of Exandria': np.array([2, 0, 1, 0, 0, 0]),
    'Leonin': np.array([1, 0, 2, 0, 0, 0]),
    'Satyr': np.array([0, 1, 0, 0, 0, 2]),
    'Fairy': np.array([0, 0, 0, 0, 0, 0]),
    'Harengon': np.array([0, 0, 0, 0, 0, 0]),
    'Aarakocra': np.array([0, 2, 0, 0, 1, 0]),
    'Genasi': np.array([0, 0, 2, 0, 0, 0]), # additional +1 based on element
    'Goliath': np.array([2, 0, 1, 0, 0, 0]),
    'Aasimar': np.array([0, 0, 0, 0, 0, 2]),
    'Bugbear': np.array([2, 1, 0, 0, 0, 0]),
    'Firbolg': np.array([1, 0, 0, 0, 2, 0]),
    'Goblin': np.array([0, 2, 1, 0, 0, 0]),
    'Hobgoblin': np.array([0, 0, 2, 1, 0, 0]),
    'Kenku': np.array([0, 2, 0, 0, 1, 0]),
    'Kobold': np.array([0, 2, 0, 0, 0, 0]),
    'Lizardfolk': np.array([0, 0, 2, 0, 1, 0]),
    'Orc': np.array([2, 0, 1, 0, 0, 0]),
    'Tabaxi': np.array([0, 2, 0, 0, 0, 1]),
    'Triton': np.array([1, 0, 1, 0, 0, 1]),
    'Yuan-Ti Pureblood': np.array([0, 0, 0, 1, 0, 2]),
    'Feral Tiefling': np.array([0, 2, 0, 1, 0, 0]),
    'Changeling': np.array([0, 0, 0, 0, 0, 2]), # choose additional +1
    'Kalashtar': np.array([0, 0, 0, 0, 2, 1]),
    'Orc of Eberron': np.array([2, 0, 1, 0, 0, 0]),
    'Shifter': np.array([0, 0, 0, 0, 0, 0]),
    'Warforged': np.array([0, 0, 2, 0, 0, 0]), # choose additional +1
    'Centaur': np.array([2, 0, 0, 0, 1, 0]),
    'Loxodon': np.array([0, 0, 2, 0, 1, 0]),
    'Minotaur': np.array([2, 0, 1, 0, 0, 0]),
    'Simic Hybrid': np.array([0, 0, 2, 0, 0, 0]), # choose additional +1
    'Vedalken': np.array([0, 0, 0, 2, 1, 0])
    }

# simulated dicerolls for base stats
def roll_4_drop_lowest():
    dicerolls = [random.randint(1,6), random.randint(1,6), random.randint(1,6), random.randint(1,6)]
    #if dicerolls.count(1) >= 2:
    #    dicerolls = [random.randint(1,6), random.randint(1,6), random.randint(1,6), random.randint(1,6)]
    dicerolls = sorted(dicerolls, reverse=True)
    dicerolls = dicerolls[0:3]
    return sum(dicerolls)

# generate stats, assigned randomly to ability scores
def generate_unweighted_stats(race_choice):
    base_stats = [roll_4_drop_lowest(), roll_4_drop_lowest(), roll_4_drop_lowest(), roll_4_drop_lowest(), roll_4_drop_lowest(), roll_4_drop_lowest()]
    total_stats = base_stats + race_bonuses[race_choice]
    total_modifiers = calc_mod_from_score(total_stats)
    return dict(zip(ability_scores, base_stats)), dict(zip(ability_scores, race_bonuses[race_choice])), dict(zip(ability_scores, total_stats)), dict(zip(ability_scores, total_modifiers))

# generate stats, assign with priority to certain scores, assign rest randomly
def generate_weighted_stats(race_choice, class_choice):
    base_stats = [0, 0, 0, 0, 0, 0]
    sorted_stats = sorted([roll_4_drop_lowest(), roll_4_drop_lowest(), roll_4_drop_lowest(), roll_4_drop_lowest(), roll_4_drop_lowest(), roll_4_drop_lowest()], reverse=True)
    if class_choice == 'Artificer':
        base_stats[3] = sorted_stats.pop(0)
        base_stats[1] = sorted_stats.pop(0)
        random.shuffle(sorted_stats)
        base_stats[0], base_stats[2], base_stats[4], base_stats[5] = sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0)
    elif class_choice == 'Barbarian':
        base_stats[0] = sorted_stats.pop(0)
        base_stats[2] = sorted_stats.pop(0)
        base_stats[1] = sorted_stats.pop(0)
        random.shuffle(sorted_stats)
        base_stats[3], base_stats[4], base_stats[5] = sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0)
    elif class_choice == 'Bard':
        base_stats[5] = sorted_stats.pop(0)
        base_stats[1] = sorted_stats.pop(0)
        random.shuffle(sorted_stats)
        base_stats[0], base_stats[2], base_stats[3], base_stats[4] = sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0)
    elif class_choice == 'Cleric':
        base_stats[4] = sorted_stats.pop(0)
        base_stats[2] = sorted_stats.pop(0)
        random.shuffle(sorted_stats)
        base_stats[0], base_stats[1], base_stats[3], base_stats[5] = sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0)
    elif class_choice == 'Druid':
        base_stats[4] = sorted_stats.pop(0)
        base_stats[2] = sorted_stats.pop(0)
        random.shuffle(sorted_stats)
        base_stats[0], base_stats[1], base_stats[3], base_stats[5] = sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0)
    elif class_choice == 'Fighter':
        base_stats[0] = sorted_stats.pop(0)
        base_stats[1] = sorted_stats.pop(0)
        base_stats[2] = sorted_stats.pop(0)
        random.shuffle(sorted_stats)
        base_stats[3], base_stats[4], base_stats[5] = sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0)
    elif class_choice == 'Monk':
        base_stats[1] = sorted_stats.pop(0)
        base_stats[4] = sorted_stats.pop(0)
        base_stats[2] = sorted_stats.pop(0)
        random.shuffle(sorted_stats)
        base_stats[0], base_stats[3], base_stats[5] = sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0)
    elif class_choice == 'Paladin':
        base_stats[5] = sorted_stats.pop(0)
        base_stats[0] = sorted_stats.pop(0)
        base_stats[3] = sorted_stats.pop(0)
        random.shuffle(sorted_stats)
        base_stats[1], base_stats[2], base_stats[4] = sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0)
    elif class_choice == 'Ranger':
        base_stats[1] = sorted_stats.pop(0)
        base_stats[4] = sorted_stats.pop(0)
        random.shuffle(sorted_stats)
        base_stats[0], base_stats[2], base_stats[3], base_stats[5] = sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0)
    elif class_choice == 'Rogue':
        base_stats[1] = sorted_stats.pop(0)
        random.shuffle(sorted_stats)
        base_stats[0], base_stats[2], base_stats[3], base_stats[4], base_stats[5] = sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0)
    elif class_choice == 'Sorcerer':
        base_stats[5] = sorted_stats.pop(0)
        base_stats[2] = sorted_stats.pop(0)
        random.shuffle(sorted_stats)
        base_stats[0], base_stats[1], base_stats[3], base_stats[4] = sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0)
    elif class_choice == 'Warlock':
        base_stats[5] = sorted_stats.pop(0)
        base_stats[1] = sorted_stats.pop(0)
        random.shuffle(sorted_stats)
        base_stats[0], base_stats[2], base_stats[3], base_stats[4] = sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0)
    elif class_choice == 'Wizard':
        base_stats[3] = sorted_stats.pop(0)
        random.shuffle(sorted_stats)
        base_stats[0], base_stats[1], base_stats[2], base_stats[4], base_stats[5] = sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0), sorted_stats.pop(0)
    total_stats = base_stats + race_bonuses[race_choice]
    total_modifiers = calc_mod_from_score(total_stats)
    return dict(zip(ability_scores, base_stats)), dict(zip(ability_scores, race_bonuses[race_choice])), dict(zip(ability_scores, total_stats)), dict(zip(ability_scores, total_modifiers))

# calculate hitpoints based on con mod, class, and char level
def calc_hitpoints(con_mod, classchoice, char_level):
    hitpoints = 0
    if classchoice == 'Artificer':
        hitpoints = (8+con_mod) + (5+con_mod)*(char_level-1)
    if classchoice == 'Barbarian':
        hitpoints = (12+con_mod) + (7+con_mod)*(char_level-1)
    if classchoice == 'Bard':
        hitpoints = (8+con_mod) + (5+con_mod)*(char_level-1)
    if classchoice == 'Cleric':
        hitpoints = (8+con_mod) + (5+con_mod)*(char_level-1)
    if classchoice == 'Druid':
        hitpoints = (8+con_mod) + (5+con_mod)*(char_level-1)
    if classchoice == 'Fighter':
        hitpoints = (10+con_mod) + (6+con_mod)*(char_level-1)
    if classchoice == 'Monk':
        hitpoints = (8+con_mod) + (5+con_mod)*(char_level-1)
    if classchoice == 'Paladin':
        hitpoints = (10+con_mod) + (6+con_mod)*(char_level-1)
    if classchoice == 'Ranger':
        hitpoints = (10+con_mod) + (6+con_mod)*(char_level-1)
    if classchoice == 'Rogue':
        hitpoints = (8+con_mod) + (5+con_mod)*(char_level-1)
    if classchoice == 'Sorcerer':
        hitpoints = (6+con_mod) + (4+con_mod)*(char_level-1)
    if classchoice == 'Warlock':
        hitpoints = (8+con_mod) + (5+con_mod)*(char_level-1)
    if classchoice == 'Wizard':
        hitpoints = (6+con_mod) + (4+con_mod)*(char_level-1)
    return str(hitpoints)

# calculate armor class based on armor available in proficiencies and equipment
def calc_armorclass(modifiers, classchoice, armor_profs, equipment_list):
    str_mod, dex_mod, con_mod = modifiers['STR'], modifiers['DEX'], modifiers['CON']
    int_mod, wis_mod, cha_mod = modifiers['INT'], modifiers['WIS'], modifiers['CHA']
    armorclass = 10+dex_mod
    if classchoice == 'Barbarian':
        armorclass = 10+dex_mod+con_mod
    elif classchoice == 'Monk':
        armorclass = 10+dex_mod+wis_mod
    else:
        if ('Chain Mail' in equipment_list) & ('Heavy Armor' in armor_profs):
            armorclass = 16
        elif ('Scale Mail' in equipment_list) & ('Medium Armor' in armor_profs):
            if dex_mod >= 2:
                armorclass = 16
            else:
                armorclass = 14+dex_mod
        elif ('Studded Leather Armor' in equipment_list) & ('Light Armor' in armor_profs):
            armorclass = 12+dex_mod
        elif ('Leather Armor' in equipment_list) & ('Light Armor' in armor_profs):
            armorclass = 11+dex_mod
    if ('Shield' in equipment_list) & ('Shields' in armor_profs):
        armorclass = armorclass+2
    return str(armorclass)

File: generateStats.js
// import lodash library
import lodash from 'lodash';

// array of ability score abbreviations
const abilityScores = ['STR', 'DEX', 'CON', 'INT', 'WIS', 'CHA']

// (k, v) pairs of (race, bonuses)
const raceBonuses = {
  'Dragonborn': [0, 2, 0, 0, 0, 1],
  'Hill Dwarf': [0, 0, 2, 0, 1, 0],
  'Mountain Dwarf': [2, 0, 2, 0, 0, 0],
  'Drow Elf': [0, 2, 0, 0, 0, 1],
  'High Elf': [0, 2, 0, 1, 0, 0],
  'Wood Elf': [0, 2, 0, 0, 1, 0],
  'Forest Gnome': [0, 1, 0, 2, 0, 0],
  'Rock Gnome': [0, 0, 1, 2, 0, 0],
  'Half-Elf': [0, 0, 0, 0, 0, 2], // two additional +1's
  'Lightfoot Halfling': [0, 2, 0, 0, 0, 1],
  'Stout Halfling': [0, 2, 1 , 0, 0, 0],
  'Half-Orc': [2, 0, 1, 0, 0, 0],
  'Human': [1, 1, 1, 1, 1, 1],
  'Variant Human': [0, 0, 0, 0, 0, 0], // choose two +1's
  'Tiefling': [0, 0, 0, 1, 0, 2],
  'Orc of Exandria': [2, 0, 1, 0, 0, 0],
  'Leonin': [1, 0, 2, 0, 0, 0],
  'Satyr': [0, 1, 0, 0, 0, 2],
  'Fairy': [0, 0, 0, 0, 0, 0],
  'Harengon': [0, 0, 0, 0, 0, 0],
  'Aarakocra': [0, 2, 0, 0, 1, 0],
  'Genasi': [0, 0, 2, 0, 0, 0], // additional +1 based on element
  'Goliath': [2, 0, 1, 0, 0, 0],
  'Aasimar': [0, 0, 0, 0, 0, 2],
  'Bugbear': [2, 1, 0, 0, 0, 0],
  'Firbolg': [1, 0, 0, 0, 2, 0],
  'Goblin': [0, 2, 1, 0, 0, 0],
  'Hobgoblin': [0, 0, 2, 1, 0, 0],
  'Kenku': [0, 2, 0, 0, 1, 0],
  'Kobold': [0, 2, 0, 0, 0, 0],
  'Lizardfolk': [0, 0, 2, 0, 1, 0],
  'Orc': [2, 0, 1, 0, 0, 0],
  'Tabaxi': [0, 2, 0, 0, 0, 1],
  'Triton': [1, 0, 1, 0, 0, 1],
  'Yuan-Ti Pureblood': [0, 0, 0, 1, 0, 2],
  'Feral Tiefling': [0, 2, 0, 1, 0, 0],
  'Changeling': [0, 0, 0, 0, 0, 2], // choose additional +1
  'Kalashtar': [0, 0, 0, 0, 2, 1],
  'Orc of Eberron': [2, 0, 1, 0, 0, 0],
  'Shifter': [0, 0, 0, 0, 0, 0],
  'Warforged': [0, 0, 2, 0, 0, 0], // choose additional +1
  'Centaur': [2, 0, 0, 0, 1, 0],
  'Loxodon': [0, 0, 2, 0, 1, 0],
  'Minotaur': [2, 0, 1, 0, 0, 0],
  'Simic Hybrid': [0, 0, 2, 0, 0, 0], // choose additional +1
  'Vedalken': [0, 0, 0, 2, 1, 0]
}

// get array of race names
const races = Object.keys(raceBonuses);

// function to calculate mod based on total score
const calcModFromScore = (abilityScore) => {
  let abilityMod = Math.floor((abilityScore-10)/2);
  return abilityMod;
}

// simulated dicerolls for base stats
const roll4DropLowest = () => {
  let dicerolls = [Math.floor(Math.random() * 6) + 1, 
                  Math.floor(Math.random() * 6) + 1, 
                  Math.floor(Math.random() * 6) + 1, 
                  Math.floor(Math.random() * 6) + 1];
  dicerolls.sort().reverse();
  return lodash.sum(dicerolls.slice(0,3));
}

// zip together ability score names and stats
const zipStats = (statArray) => {
  let zippedArray = statArray.reduce((result, stat, index) => {
    result[abilityScores[index]] = stat;
    return result;
  }, {})
  return zippedArray;
}

// generate stats, assigned randomly to ability scores
const generateUnweightedStats = (raceChoice) => {
  console.log(''); 
  console.log(raceChoice); 
  console.log('');
  let baseStats = [roll4DropLowest(), roll4DropLowest(), roll4DropLowest(), roll4DropLowest(), roll4DropLowest(), roll4DropLowest()];
  let totalStats = lodash.zipWith(baseStats, raceBonuses[raceChoice], lodash.add);
  let totalModifiers = totalStats.map(i => calcModFromScore(i));
  let statsObject = {'Base Stats': zipStats(baseStats), 'Total Stats': zipStats(totalStats), 'Total Modifiers': zipStats(totalModifiers)};
  return statsObject;
}

// generate stats, assign with priority to certain scores, assign rest randomly
const generateWeightedStats = (raceChoice, classChoice) => {
  console.log(''); 
  console.log(raceChoice); 
  console.log(classChoice); 
  console.log('');
  let baseStats = [0, 0, 0, 0, 0, 0];
  let sortedStats = [roll4DropLowest(), 
                roll4DropLowest(), 
                roll4DropLowest(), 
                roll4DropLowest(), 
                roll4DropLowest(), 
                roll4DropLowest()];
  sortedStats.sort((a,b) => {a-b}).reverse();
  switch (classChoice) {
    case 'Artificer':
      baseStats[3] = sortedStats.shift();
      baseStats[1] = sortedStats.shift();
      baseStats[0] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[2] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[4] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[5] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0];
      break;
    case 'Barbarian':
      baseStats[0] = sortedStats.shift();
      baseStats[2] = sortedStats.shift(); 
      baseStats[1] = sortedStats.shift(); 
      baseStats[3] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[4] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[5] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0];
      break; 
    case 'Bard':
      baseStats[5] = sortedStats.shift(); 
      baseStats[1] = sortedStats.shift(); 
      baseStats[0] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[2] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[3] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[4] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0];
      break; 
    case 'Cleric':
      baseStats[4] = sortedStats.shift(); 
      baseStats[2] = sortedStats.shift(); 
      baseStats[0] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[1] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[3] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[5] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      break;
    case 'Druid':
      baseStats[4] = sortedStats.shift(); 
      baseStats[2] = sortedStats.shift(); 
      baseStats[0] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[1] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[3] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[5] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0];
      break; 
    case 'Fighter':
      baseStats[0] = sortedStats.shift(); 
      baseStats[1] = sortedStats.shift(); 
      baseStats[2] = sortedStats.shift(); 
      baseStats[3] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[4] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[5] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0];
      break;
    case 'Monk':
      baseStats[1] = sortedStats.shift(); 
      baseStats[4] = sortedStats.shift(); 
      baseStats[2] = sortedStats.shift(); 
      baseStats[0] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[3] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[5] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0];
      break; 
    case 'Paladin':
      baseStats[5] = sortedStats.shift(); 
      baseStats[0] = sortedStats.shift(); 
      baseStats[3] = sortedStats.shift(); 
      baseStats[1] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[2] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[4] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0];
      break; 
    case 'Ranger':
      baseStats[1] = sortedStats.shift(); 
      baseStats[4] = sortedStats.shift(); 
      baseStats[0] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[2] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[3] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[5] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0];
      break; 
    case 'Rogue':
      baseStats[1] = sortedStats.shift(); 
      baseStats[0] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[2] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[3] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[4] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[5] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0];
      break; 
    case 'Sorcerer':
      baseStats[5] = sortedStats.shift(); 
      baseStats[2] = sortedStats.shift(); 
      baseStats[0] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[1] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[3] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[4] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      break;
    case 'Warlock':
      baseStats[5] = sortedStats.shift(); 
      baseStats[1] = sortedStats.shift(); 
      baseStats[0] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[2] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[3] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[4] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0];
      break; 
    case 'Wizard':
      baseStats[3] = sortedStats.shift(); 
      baseStats[0] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[1] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[2] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[4] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0]; 
      baseStats[5] = sortedStats.splice(Math.floor(Math.random() * sortedStats.length), 1)[0];
      break; 
    default:
      break;
  }
  let totalStats = lodash.zipWith(baseStats, raceBonuses[raceChoice], lodash.add);
  let totalModifiers = totalStats.map(i => calcModFromScore(i));
  let statsObject = {'Base Stats': zipStats(baseStats), 
                'Total Stats': zipStats(totalStats), 
                'Total Modifiers': zipStats(totalModifiers)};
  return statsObject;
}

// calculate hitpoints based on con mod, class, and char level
const calcHitpoints = (conMod, classChoice, charLevel) => {
  let hitpoints = 0;
  switch (classChoice){
    case 'Artificer':
    case 'Bard':
    case 'Cleric':
    case 'Druid':
    case 'Monk':
    case 'Rogue':
    case 'Warlock':
      hitpoints = (8 + conMod) + (5 + conMod) * (charLevel - 1);
      break;
    case 'Barbarian':
      hitpoints = (12 + conMod) + (8 + conMod) * (charLevel - 1);
      break;
    case 'Fighter':
    case 'Paladin':
    case 'Ranger':
      hitpoints = (10 + conMod) + (6 + conMod) * (charLevel - 1);
      break;
    case 'Sorcerer':
    case 'Wizard':
      hitpoints = (6 + conMod) + (4 + conMod) * (charLevel - 1);
      break;
    default:
      break;
  }
  return hitpoints;
}

// calculate armor class based on available gear and proficiencies
const calcArmorClass = (modifiers, classChoice, equipmentList) => {
  modSTR = modifiers['STR'];
  modDEX = modifiers['DEX'];
  modCON = modifiers['CON'];
  modINT = modifiers['INT'];
  modWIS = modifiers['WIS'];
  modCHA = modifiers['CHA'];
  armorClass = 10 + modDEX
  // special cases of Barbarian and Monk
  if (classChoice === 'Barbarian') {
    armorClass = 10 + modDEX + modSTR 
  } else if (classChoice === 'Monk') {
    armorClass = 10 + modDEX + modWIS
  }
  // normal cases of armor in inventory
  if (equipmentList.includes('Chain Mail')) {
    armorClass = 16
  } else if (equipmentList.includes('Scale Mail')) {
    if (modDEX >= 2) {
      armorClass = 16
    } else {
      armorClass = 14 + modDEX
    }
  } else if (equipmentList.includes('Studded Leather Armor')) {
    armorClass = 12 + modDEX
  } else if (equipmentList.includes('Leather Armor')) {
    armorClass = 11 + modDEX
  }
  if (equipmentList.includes('Shield')) {
    armorClass = armorClass + 2
  }
  return armorClass
}

export { generateWeightedStats, generateUnweightedStats, calcModFromScore, calcArmorClass, calcHitpoints };
// console.log('Unweighted: ')
// console.log(generateUnweightedStats(races[Math.floor(Math.random()*races.length)]));
// console.log('')
// console.log('Weighted: ')
// console.log(generateWeightedStats(races[Math.floor(Math.random()*races.length)], 'Fighter'));